eyeball-im-util/src/vector/sort.rs

```diff
@@ -0,0 +1,613 @@
+    /// // We get a `VectorDiff::Append` with sorted values!
+    /// assert_next_eq!(sub, VectorDiff::Append { values: vector!['b', 'd', 'e'] });
+    ///
+    /// // Let's recap what we have. `ob` is our `ObservableVector`,
+    /// // `sub` is the “sorted view”/“sorted stream” of `ob`:
+    /// // | `ob`  | d b e |
+    /// // | `sub` | b d e |
+    ///
+    /// // Append other multiple values.
+    /// ob.append(vector!['f', 'g', 'a', 'c']);
+    /// // We get three `VectorDiff`s!
+    /// assert_next_eq!(sub, VectorDiff::PushFront { value: 'a' });
+    /// assert_next_eq!(sub, VectorDiff::Insert { index: 2, value: 'c' });
+    /// assert_next_eq!(sub, VectorDiff::Append { values: vector!['f', 'g'] });
+    ///
+    /// // Let's recap what we have:
+    /// // | `ob`  | d b e f g a c |
+    /// // | `sub` | a b c d e f g |
+    /// //           ^   ^     ^^^
+    /// //           |   |     |
+    /// //           |   |     with `VectorDiff::Append { .. }`
+    /// //           |   with `VectorDiff::Insert { index: 2, .. }`
+    /// //           with `VectorDiff::PushFront { .. }`
+    ///
+    /// // Technically, `Sort` emits `VectorDiff`s that mimic a sorted `Vector`.
+    ///
+    /// drop(ob);
+    /// assert_closed!(sub);
+    /// # }
+    /// ```
+    ///
+    /// [`ObservableVector`]: eyeball_im::ObservableVector
+    #[project = SortProj]
+    pub struct Sort<S, F>
+    where
+        S: Stream,
+        S::Item: VectorDiffContainer,
+    {
+        // The main stream to poll items from.
+        #[pin]
+        inner_stream: S,
+
+        // The comparison function to sort items.
+        compare: F,
+
+        // This is the **sorted** buffered vector.
+        buffered_vector: Vector<(UnsortedIndex, VectorDiffContainerStreamElement<S>)>,
+
+        // This adapter can produce many items per item of the underlying stream.
+        //
+        // Thus, if the item type is just `VectorDiff<_>` (non-bached, can't
+        // just add diffs to a `poll_next` result), we need a buffer to store the
+        // possible extra items in.
+        ready_values: VectorDiffContainerStreamBuffer<S>,
+    }
+}
+
+impl<S, F> Sort<S, F>
+where
+    S: Stream,
+    S::Item: VectorDiffContainer,
+    F: Fn(&VectorDiffContainerStreamElement<S>, &VectorDiffContainerStreamElement<S>) -> Ordering
+        + Clone,
+{
+    /// Create a new `Filter` with the given (unfiltered) initial values, stream
+    /// of `VectorDiff` updates for those values, and filter.
+    pub fn new(
+        initial_values: Vector<VectorDiffContainerStreamElement<S>>,
+        inner_stream: S,
+        compare: F,
+    ) -> (Vector<VectorDiffContainerStreamElement<S>>, Self) {
+        let mut initial_values = initial_values.into_iter().enumerate().collect::<Vector<_>>();
+        initial_values.sort_by(|(_, left), (_, right)| compare(left, right));
+
+        (
+            initial_values.clone().into_iter().map(|(_, value)| value).collect(),
```

Comment:
Since there is no way around every value here anyways, this should be more efficient:

```suggestion
            initial_values.iter().map(|(_, value)| value.clone()).collect(),
```